mountpoint-s3-client/src/s3_crt_client.rs

```diff
@@ -976,8 +1007,15 @@ fn try_parse_generic_error(request_result: &MetaRequestResult) -> Option<S3Reque
             let message = error_elem
                 .get_child("Message")
                 .and_then(|e| e.get_text())
-                .unwrap_or(error_code_str);
-            Some(S3RequestError::Forbidden(message.into_owned()))
+                .unwrap_or(error_code_str.clone());
+            Some(S3RequestError::Forbidden(
+                message.to_string(),
+                Box::new(ClientErrorMetadata {
```

Comment:
One aspect of this change that does not convince me completely is that most (all?) metadata information seems to be "intrinsic" to the type of error rather than additional. This here is the only case where we add the code/message, but it is largely duplicated.
Are there other cases to be implemented in the future that drive your approach?
Otherwise, do we really need to add the (boxed) metadata struct, or could we build it on the fly based on the enum case (giving up on the return reference)?

Reply:
in the [last commit](https://github.com/awslabs/mountpoint-s3/pull/882/commits/2ec7627756199a0a1426185fa497c013f51cb24c) I've changed this to be constructed in `ProvideErrorMetadata::meta()` for `ResponseError` and `Throttled`, which required changing signature to `fn meta(&self) -> ClientErrorMetadata` (no ref in returned value)

I like the fact that this logic is grouped in `ProvideErrorMetadata` trait implementation for `S3RequestError` and that now we can use `Default::default()` instead of `&ClientErrorMetadata::EMPTY`

ps. we make clients of `mountpoint-s3-client` to always copy metadata even if a reference could've suffice, but it does not change current usages and I think that is a reasonable trade-off to make to have a more clear code